### PR TITLE
Make -inits not destroy its argument

### LINKED ISSUE #312: -inits destroys inputs
The list passed to `-inits` is destroyed by the `nreverse`. Here's a version that wouldn't (and is lower time complexity, I think.)

```elisp
(defun -inits (list)
  "Return all prefixes of LIST."
  (let ((acc nil)
        (ret (list nil)))
    (--each list
      (!cons it acc)
      (!cons (reverse acc) ret))
    (nreverse ret)))
```

### PR DESCRIPTION
**Pull request by SwiftLawnGnome**
Fixes #312

### CODE REVIEW

File: dash.el [hunk 1]
```diff
@@ -2342,7 +2342,12 @@ or with `-compare-fn' if that's non-nil."
 
 (defun -inits (list)
   "Return all prefixes of LIST."
-  (nreverse (-map 'reverse (-tails (nreverse list)))))
+  (let ((acc nil)
+        (ret (list nil)))
+    (--each list
+      (!cons it acc)
+      (!cons (reverse acc) ret))
```
Comment (basil-conto) [inline dash.el]: Nitpick: Where equally convenient, prefer built-ins (`push`) over their Dash counterparts (`!cons`).

**Comment by basil-conto** [comment]
The implementation using `--each` that you suggest is indeed more efficient than the incumbent one, and is almost as efficient as I think this can get in Elisp. Following some benchmarking, however, I think I was able to improve on it slightly. First of all, the most succinct and functional implementation:

```el
(defun -inits (list)
  (--reductions-from (append acc (list it)) () list))
```

where `--reductions-from` is a (currently non-existent) anaphoric version of (the existing) `-reductions-from`:

```el
(defmacro --reductions-from (form init list)
  `(nreverse (--reduce-from (cons (let ((acc (car acc))) ,form) acc)
                            (list ,init) ,list)))

(defun -reductions-from (fn init list)
  (--reductions-from (funcall fn acc it) init list))
```

To my pleasant surprise, using `--reductions-from` exhibits nearly identical performance to the suggested `--each` implementation.

Next, the most efficient implementation I was able to come up with:

```el
(defun -inits (list)
  (let ((res (list list)))
    (setq list (reverse list))
    (while list
      (push (reverse (!cdr list)) res))
    res))
```

Here are the various benchmarks I tried:

```el
;;; inits.el --- benchmarks for -inits -*- lexical-binding: t -*-

(require 'dash)

(defun inits-tails (list)
  (nreverse (mapcar #'reverse (-tails (reverse list)))))

(defun inits-cdr (list)
  (let ((res (list list)))
    (setq list (reverse list))
    (while list
      (push (reverse (!cdr list)) res))
    res))

(defun inits-each (list)
  (let ((ret (list ()))
        acc)
    (--each list
      (push it acc)
      (push (reverse acc) ret))
    (nreverse ret)))

(defun inits-each-map (list)
  (let ((ret (list ()))
        acc)
    (--each list
      (push it acc)
      (push acc ret))
    (nreverse (mapcar #'reverse ret))))

(defun inits-butlast (list)
  (let ((res (list list)))
    (while list
      (let ((bl (butlast list)))
        (push bl res)
        (setq list bl)))
    res))

(defun inits-reduct (list)
  (--reductions-from (append acc (list it)) () list))

(defun inits-map (list)
  (let (acc)
    (cons () (--map (reverse (push it acc)) list))))

(defun inits-append (list)
  (let (acc)
    (cons () (--map (setq acc (append acc (list it))) list))))

(defun inits-vec (list)
  (let ((vec (vconcat list)))
    (cons () (--map-indexed (append (substring vec 0 (1+ it-index)) ())
                            list))))

(let ((l (number-sequence 0 511)))
  (dolist (fn '(inits-tails
                inits-tails
                inits-cdr
                inits-each
                inits-each-map
                inits-butlast
                inits-reduct
                inits-map
                inits-append
                inits-vec))
    (garbage-collect)
    (apply #'message "%-15s %.3f %d %.3f" fn
           (benchmark-run 100
             (funcall fn l)))))

;;; inits.el ends here
```

Running them as:

```sh
emacs -batch -L /path/to/dash -f batch-byte-compile inits.el \
  && emacs -batch -L /path/to/dash -l inits.elc
```

in the following Emacs build:

```
In GNU Emacs 27.0.50 (build 1, x86_64-pc-linux-gnu, X toolkit, Xaw3d scroll bars)
 of 2019-09-18 built on thunk
Repository revision: ae3edf0ac3f1e893338917497b55859d6aca7d42
Repository branch: master
Windowing system distributor 'The X.Org Foundation', version 11.0.12004000
System Description: Debian GNU/Linux bullseye/sid

Configured using:
 'configure 'CC=ccache gcc' 'CFLAGS=-O2 -march=native' --config-cache
 --prefix=/home/blc/.local --with-mailutils --with-x-toolkit=lucid
 --with-modules --with-file-notification=yes --with-x'

Configured features:
XAW3D XPM JPEG TIFF GIF PNG RSVG SOUND GPM DBUS GSETTINGS GLIB NOTIFY
INOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF
XFT ZLIB TOOLKIT_SCROLL_BARS LUCID X11 XDBE XIM MODULES THREADS
LIBSYSTEMD JSON PDUMPER LCMS2 GMP
```

gave the following results:

```
inits-tails     2.684 267 2.554
inits-tails     2.730 266 2.613
inits-cdr       2.561 262 2.449
inits-each      2.606 263 2.459
inits-each-map  2.814 264 2.719
inits-butlast   2.676 262 2.460
inits-reduct    2.609 262 2.446
inits-map       2.867 263 2.728
inits-append    2.818 262 2.667
inits-vec       3.839 394 3.657
```

As such, my first preference would be to go with `inits-cdr` for performance, followed by `inits-reduct` for conciseness, and only then for the suggested `inits-each`. WDYT?

**Comment by basil-conto** [comment]
> where `--reductions-from` is a (currently non-existent) anaphoric version of (the existing) `-reductions-from`

I have now submitted draft PR #314 which adds this and other related anaphoric reductions macros.

**Comment by Fuco1** [comment]
The performance characteristics also probably depend on your given CPU and things.  The differences are in the range of +/- 5% which to me is not enough to decide purely on that point.

Having said that, for the implementation we can use the `cdr` one, it's simple to understand and debug.  The current version of inits simply exists because it took me 10 seconds to implement it, I don't particularly care to have things in functional style as pretty much no user of the library cares how things work under the hood.

**Comment by Fuco1** [comment]
@basil-conto You've created a lot of pull requests now, which is awesome!  I'm getting a bit lost in the dependencies between them (which there seem to be some to me).  Can you write out what needs to be merged in what order?

Alternatively, we can promote you to collaborator, the patches are good quality, with benchmarks and tests.  Makes sense to me. @magnars what do you think?

**Comment by magnars** [comment]
Thumbs for up making @basil-conto a collaborator.

**Comment by basil-conto** [comment]
> The performance characteristics also probably depend on your given CPU and things.

Of course, though I think the Emacs build configuration makes a bigger difference there.

> The differences are in the range of +/- 5% which to me is not enough to decide purely on that point.

I have a feeling that the differences increase with larger input lists.

> Having said that, for the implementation we can use the cdr one, it's simple to understand and debug.

SGTM.

> I don't particularly care to have things in functional style as pretty much no user of the library cares how things work under the hood.

Indeed it is better to provide a functional API than a functional implementation. Performance matters for clients.

> You've created a lot of pull requests now, which is awesome! I'm getting a bit lost in the dependencies between them (which there seem to be some to me). Can you write out what needs to be merged in what order?

Sorry about the deluge! There's only one dependency for my PRs: #309. That PR fixes the test suite so that it can run on all Emacs versions 24-27.

This is why I have marked my other PRs as drafts (i.e. not ready to be reviewed/merged) until #309 is fixed. Once it is merged, the other PRs are pretty independent of one another (except for the fact that the docs need regenerating after each modification).

> Alternatively, we can promote you to collaborator

I would be happy and honoured to take on this responsibility, though I would obviously still seek discussion and approval before making any significant changes. Thanks!

**Comment by basil-conto** [comment]
>> Having said that, for the implementation we can use the cdr one, it's simple to understand and debug.

> SGTM.

If you also agree @SwiftLawnGnome, feel free to update your PR accordingly. Since this involves a change of fewer than 15 LoC, it counts as a "tiny change" and can be merged even before your copyright assignment is on file.

**Comment by SwiftLawnGnome** [comment]
Agreed, I just committed @basil-conto's implementation

**Comment by Fuco1** [comment]
Thanks everyone!

**Pull request merged**